### PR TITLE
Implement with_language() and local_language()

### PR DESCRIPTION
**Pull request by hadley**
WIP: want to see if these tests fail due to gettext caching

### CODE REVIEW

**Comment by hadley** [comment]
@MichaelChirico looks like I figured out a way to clear the cache here too

File: R/language.R [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#' Langugage
```
Comment (MichaelChirico) [inline R/language.R:1]: typo

File: R/language.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#'
+#' Temporarily change language used for translations.
+#'
+#' @param lang A BCP47 language code like "en" (English), "fr" (French),
+#'   "fr_CA" (French Canadian).
+#' @inheritParams with_collate
+#' @export
+#' @examples
+#'
+#' with_language("en", try(mean[[1]]))
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/language.R:11]: Equally inscrutable in every language 🤣

File: R/language.R [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#' @rdname with_language
+local_language <- function(lang, .local_envir = parent.frame()) {
+  if (identical(Sys.getenv("LANG"), "C")) {
+    warning("Can't change language when envvar LANG='C'")
+  }
+
+  local_envvar(LANGUAGE = lang, .local_envir = .local_envir)
+  if (Sys.info()[["sysname"]] != "Windows") {
+    # Set locale to reset cache
+    local_locale(c(LC_MESSAGES = ""), .local_envir = .local_envir)
```
Comment (MichaelChirico) [inline R/language.R:30]: interesting! was this by trial&error or did you track this down to somewhere source code?

File: R/language.R [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#'
+with_language <- function(lang, code) {
+  local_language(lang)
+  code
+}
+
+#' @export
+#' @rdname with_language
+local_language <- function(lang, .local_envir = parent.frame()) {
+  if (identical(Sys.getenv("LANG"), "C")) {
```
Comment (MichaelChirico) [inline R/language.R:23]: is this a documented thing? Im curious why. maybe worth a comment here?

**Review by MichaelChirico — APPROVED** [review]
only minor changes

File: tests/testthat/test-language.R [hunk 5]
```diff
@@ -0,0 +1,10 @@
+test_that("can temporary change language", {
```
Comment (MichaelChirico) [inline tests/testthat/test-language.R:1]: I'm AFK so can't check -- does capabilities() tell us whether translation is enabled? if so we should skip_if_not(). and probably add a similar warning  in the function itself

File: R/language.R (hunk 4 above)
Comment (hadley) [reply R/language.R:23]: I documented it in the commit message but a comment would be better.

File: tests/testthat/test-language.R (hunk 5 above)
Comment (jimhester) [reply tests/testthat/test-language.R:1]: I guess this would be `capabilities("NLS")`

Comment (hadley) [reply tests/testthat/test-language.R:1]: But are there cases that we expect this not to be available?

File: R/language.R (hunk 3 above)
Comment (hadley) [reply R/language.R:30]: It was inspired by https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=931456, leading me to look for places that might conceivable call `setlocale()`. The `""` was a guess, but after the fact I read `man setlocale(3)` and saw:

> If locale is an empty string, "", each part of the locale that
       should be modified is set according to the environment variables.
       The details are implementation-dependent.  For glibc, first
       (regardless of category), the environment variable LC_ALL is
       inspected, next the environment variable with the same name as
       the category (see the table above), and finally the environment
       variable LANG.  The first existing environment variable is used.
       If its value is not a valid locale specification, the locale is
       unchanged, and setlocale() returns NULL.

So not 100% clear that will always have the desired effect, but seems like it should usually be useful.

Comment (hadley) [reply R/language.R:39]: Also recorded this in the code so we don't have to rediscover in the future

File: tests/testthat/test-language.R (hunk 5 above)
Comment (MichaelChirico) [reply tests/testthat/test-language.R:1]: im not sure how common NLS-less installations are in the wild (I don't think there are any on CRAN e.g.) but it's one of the "standard" things you can turn on/off in a build config so I have to imagine withr will find one eventually. OTOH I would expect those users with it disabled wouldn't be trying to use this in the first place... but the check seems simple enough that erring on caution makes sense to me

Comment (hadley) [reply tests/testthat/test-language.R:1]: Yeah, it's at least worth a warning here to hopefully avoid mysterious situations.

**Comment by hadley** [comment]
Also related is https://bugs.r-project.org/show_bug.cgi?id=18055, where Martin Maechler solves the problem using  `bindtextdomain()`.

**Review by jimhester — APPROVED** [review]

**Pull request merged**